File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/model/GuidedNavigationDocument.swift
import Foundation
import ReadiumShared

/// A Readium Guided Navigation Document, describing a structured sequence of
/// media-aligned navigation steps for a publication.
///
/// See https://readium.org/guided-navigation/schema/document.schema.json
struct GuidedNavigationDocument: Equatable {
  /// Optional cross-references to related resources, using the Readium Web
  /// Publication Manifest link schema.
  let links: [Link]

  /// The ordered list of guided navigation objects. Contains at least one entry.
  let guided: [GuidedNavigationObject]

  func toJson() -> [String: Any] {
    var res: [String: Any] = [:]
    if !links.isEmpty { res["links"] = links.map { $0.json } }
    res["guided"] = guided.map { $0.toJson() }
    return res
  }

  func toJsonString(pretty: Bool = false) -> String? {
    let options: JSONSerialization.WritingOptions = pretty ? [.prettyPrinted] : []
    guard let data = try? JSONSerialization.data(withJSONObject: toJson(), options: options) else {
      return nil
    }
    return String(data: data, encoding: .utf8)
  }

  static func fromJson(_ json: [String: Any]?) -> GuidedNavigationDocument? {
    guard let json = json else { return nil }

    let linksJson = json["links"] as? [[String: Any]] ?? []
    let links = linksJson.compactMap { Link(json: $0) }

    let guidedJson = json["guided"] as? [[String: Any]] ?? []
    let guided = guidedJson.compactMap { GuidedNavigationObject.fromJson($0) }

    guard !guided.isEmpty else { return nil }

    return GuidedNavigationDocument(links: links, guided: guided)
  }

  static func fromJson(_ jsonString: String) throws -> GuidedNavigationDocument? {
    let data = Data(jsonString.utf8)
    let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
    return fromJson(object)
  }
}

File: flutter_readium_platform_interface/lib/src/shared/guided_navigation/guided_navigation_document.dart
import 'package:equatable/equatable.dart';
import 'package:meta/meta.dart';

import '../../utils/index.dart';
import '../publication/link.dart';
import 'guided_navigation_object.dart';

/// A Readium Guided Navigation Document, describing a structured sequence of
/// media-aligned navigation steps for a publication.
///
/// See https://readium.org/guided-navigation/schema/document.schema.json
@immutable
class GuidedNavigationDocument with Equatable implements JSONable {
  const GuidedNavigationDocument({required this.guided, this.links = const []});

  /// Optional cross-references to related resources, using the Readium Web
  /// Publication Manifest link schema.
  final List<Link> links;

  /// The ordered list of guided navigation objects. Must contain at least one.
  final List<GuidedNavigationObject> guided;

  @override
  List<Object> get props => [links, guided];

  @override
  Map<String, dynamic> toJson() => {}
    ..putIterableIfNotEmpty('links', links)
    ..putIterableIfNotEmpty('guided', guided);

  static GuidedNavigationDocument? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;

    final jsonObject = Map<String, dynamic>.of(json);
    final links = Link.fromJsonArray(
      jsonObject.optJsonArray('links', remove: true),
    );
    final guided = GuidedNavigationObject.fromJsonArray(
      jsonObject.optJsonArray('guided', remove: true),
    );

    if (guided.isEmpty) {
      ReadiumLog.d(
        'GuidedNavigationDocument: [guided] is required and must not be empty',
      );
      return null;
    }

    return GuidedNavigationDocument(links: links, guided: guided);
  }
}
